src/main/java/com/google/api/codegen/discovery/DefaultString.java

```diff
@@ -0,0 +1,150 @@
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.api.codegen.discovery;
+
+import com.google.api.codegen.Inflector;
+import com.google.auto.value.AutoValue;
+import com.google.common.collect.ImmutableList;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import javax.annotation.Nullable;
+
+/**
+ * Creates default string from path patterns.
+ */
+public class DefaultString {
+
+  private static final String WILDCARD_PATTERN = "[^/]*";
+
+  /**
+   * Returns a default string from `pattern`, or null if the pattern is not supported.
+   */
+  public static String of(String pattern) {
+    // We only care about patterns that look like
+    //  ^projects/[^/]*/subscriptions/[^/]*$
+    // Ignore if what we get looks nothing like this.
+    if (!pattern.startsWith("^")
+        || !pattern.endsWith("$")
+        || substrCount(pattern, "/") != substrCount(pattern, WILDCARD_PATTERN) * 3 - 1) {
+      return null;
+    }
+    pattern = pattern.substring(1, pattern.length() - 1);
+    ImmutableList<Elem> elems = parse(pattern);
+    if (!validElems(elems)) {
+      return null;
+    }
+
+    StringBuilder ret = new StringBuilder();
+    for (int i = 0; i < elems.size(); i += 2) {
+      String literal = elems.get(i).getLiteral();
+      ret.append('/')
+          .append(literal)
+          .append("/{MY-")
+          .append(Inflector.singularize(literal).toUpperCase())
+          .append('}');
+    }
+    return ret.substring(1);
+  }
+
+  /**
+   * Counts the number of non-overlapping instances of `needle` in `haystack`.
+   */
+  private static int substrCount(String haystack, String needle) {
+    int count = 0;
+    int fromIndex = 0;
+    for (; ; ) {
+      int index = haystack.indexOf(needle, fromIndex);
+      if (index < 0) {
+        return count;
+      }
+      fromIndex = index + needle.length();
+      count++;
+    }
+  }
+
+  /**
+   * Parses pattern, with the leading '^' and trailing '$' removed, into a list representing the
+   * pattern.
+   */
+  private static ImmutableList<Elem> parse(String pattern) {
+    List<Elem> elems = new ArrayList<>();
+    while (pattern.length() > 0) {
+      int slash;
+      if (pattern.startsWith(WILDCARD_PATTERN)) {
+        elems.add(Elem.WILDCARD);
+        pattern = pattern.substring(WILDCARD_PATTERN.length());
+      } else if ((slash = pattern.indexOf("/")) >= 0) {
+        elems.add(Elem.createLiteral(pattern.substring(0, slash)));
+        pattern = pattern.substring(slash);
+      } else {
+        elems.add(Elem.createLiteral(pattern));
+        pattern = "";
+      }
+
+      if (pattern.startsWith("/")) {
```

Comment:
I don't think this is right. The "if" is used to clean up the leading slash (if one exists) so that the next iteration can directly begin at the wildcard or literal. It might need to run following both "startsWith(WILDCARD)" or "indexOf(/) >= 0". Am I making any sense?